test/rosdep_repo_check/alpine.py

```diff
@@ -0,0 +1,118 @@
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import os
+import tarfile
+from io import TextIOWrapper
+
+from . import open_gz_url
+from . import PackageEntry
+from . import RepositoryCacheCollection
+
+
+def replace_tokens(string, os_code_name, os_arch):
+    """Replace tokens in the repository base URL."""
+    for key, value in {
+        '$basearch': os_arch,
+        '$releasever': os_code_name,
+    }.items():
+        string = string.replace(key, value)
+    return string
+
+
+def parse_apkindex(f):
+    # An example of an APKINDEX entry. Entries are divided by a blank line.
+
+    # V:2.1.1-r0
+    # A:x86_64
+    # S:6645
+    # I:28672
+    # T:RTP proxy (documentation)
+    # U:https://www.rtpproxy.org/
+    # L:BSD-2-CLause
+    # o:rtpproxy
+    # m:Natanael Copa <ncopa@alpinelinux.org>
+    # t:1602354892
+    # c:183e99f73bb1223768aa7231a836a3e98e94c03e
+    # i:docs rtpproxy=2.1.1-r0
+
+    while True:
+        entry = {}
+        while (l := f.readline().decode('utf-8')) not in ['', '\n']:
+            k, v = l.strip().split(':', 1)
+            entry[k] = v
+
+        if entry:
+            yield entry
+        else:
+            break
+
+
+def enumerate_alpine_packages(base_url, os_name, os_code_name, os_arch):
+    """
+    Enumerate packages in an Alpine repository.
+
+    :param base_url: the Alpine repository base URL.
+    :param os_name: the name of the OS associated with the repository.
+    :param os_code_name: the OS version associated with the repository.
+    :param os_arch: the system architecture associated with the repository.
+
+    :returns: an enumeration of package entries.
+    """
+
+    base_url = replace_tokens(base_url, os_code_name, os_arch)
+    apkindex_url = os.path.join(base_url, 'APKINDEX.tar.gz')
+    print('Reading Alpine package metadata from ' + apkindex_url)
+
+    with open_gz_url(apkindex_url) as f:
+        with tarfile.open(mode='r|', fileobj=f) as tf:
+            index = None
+            for ti in tf:
+                if ti.name == 'APKINDEX':
+                    index = tf.extractfile(ti)
+                    break
+            if index is None:
+                raise RuntimeError('APKINDEX url did not contain an APKINDEX file')
+
+            for index_entry in parse_apkindex(index):
+                pkg_name, pkg_version = index_entry['o'], index_entry['V']
+                pkg_filename = '%s-%s.apk' % (pkg_name, pkg_version)
+                pkg_url = os.path.join(base_url, pkg_filename)
+                yield PackageEntry(pkg_name, pkg_version, pkg_url)
```

Comment:
Earlier I said:
> ... the `PackageEntry` field `source_name` ... should represent the name of the source package that provides the given binary package. This may very well not be a relevant concept on Alpine...

I think that this concept actually does carry over to Alpine as the "origin" field. Actually, I think that the code presented here is only enumerating the "base" packages and is missing all of the "sub-packages".

It's worth noting that Alpine's packaging dashboard seems to work the same given the names of sub-packages, where Fedora's requires a source package name, so we don't actually need to populate the source package name here for things to work for Alpine. It should probably be present for completeness and future-proofing, though.

Even if we didn't populate the `source_name` field, we should definitely enumerate sub-packages since a rosdep rule could certainly refer to one.

```suggestion
                pkg_name, pkg_version, source_name = index_entry['P'], index_entry['V'], index_entry['o']
                pkg_filename = '%s-%s.apk' % (pkg_name, pkg_version)
                pkg_url = os.path.join(base_url, pkg_filename)
                yield PackageEntry(pkg_name, pkg_version, pkg_url, source_name=source_name)
```